mountpoint-s3/src/fuse/session.rs

```diff
@@ -101,3 +88,249 @@ enum Message {
+    "unknown thread id".to_string()
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::sync::{
+        mpsc::{self, Receiver},
+        Condvar, Mutex,
+    };
+    use std::time::Duration;
+    use test_case::test_case;
+
+    use super::*;
+
+    struct TestMessage {
+        _id: usize,
+        mutex: Mutex<bool>,
+        cond: Condvar,
+    }
+
+    impl TestMessage {
+        fn new(_id: usize) -> Self {
+            Self {
+                _id,
+                mutex: Mutex::new(false),
+                cond: Condvar::new(),
+            }
+        }
+
+        fn process(&self) {
+            let mut done = self.mutex.lock().unwrap();
+            while !*done {
+                done = self.cond.wait(done).unwrap();
+            }
+        }
+
+        fn complete(&self) {
+            let mut done = self.mutex.lock().unwrap();
+            *done = true;
+            self.cond.notify_one();
+        }
+    }
+    struct TestWork {
+        receiver: Arc<Mutex<Receiver<Arc<TestMessage>>>>,
+    }
+
+    impl Work for TestWork {
+        type Result = ();
+
+        fn run<FB, FA>(&self, mut before: FB, mut after: FA) -> Self::Result
+        where
+            FB: FnMut(),
+            FA: FnMut(),
+        {
+            while let Ok(message) = {
+                let receiver = self.receiver.lock().unwrap();
+                receiver.recv()
+            } {
+                before();
+                message.process();
+                after();
+            }
+        }
+    }
+
+    #[test_case(10, 10)]
+    #[test_case(10, 30)]
+    #[test_case(30, 10)]
+    fn test_worker_pool(max_worker_threads: usize, concurrent_messages: usize) {
+        let (tx, rx) = mpsc::channel();
+        let work = TestWork {
+            receiver: Arc::new(Mutex::new(rx)),
+        };
+
+        let (workers_tx, workers_rx) = mpsc::channel::<JoinHandle<()>>();
+        WorkerPool::start(work, workers_tx, max_worker_threads).unwrap();
+
+        // Send messages
+        let messages = (0..concurrent_messages)
+            .map(|i| {
+                let message = Arc::new(TestMessage::new(i));
+                tx.send(message.clone()).unwrap();
+                message
+            })
+            .collect::<Vec<_>>();
+
+        let mut workers = Vec::new();
+
+        let min_expected_workers = concurrent_messages.min(max_worker_threads);
+        for _ in 0..min_expected_workers {
+            let worker = workers_rx.recv_timeout(Duration::from_secs(1)).unwrap();
+            workers.push(worker);
+        }
+
+        // Complete messages
+        for m in messages {
+            m.complete();
+        }
+
+        drop(tx);
+
+        // The pool tries to spawn an extra idle worker.
+        if let Ok(worker) = workers_rx.recv() {
+            workers.push(worker);
+            assert_eq!(workers.len(), min_expected_workers + 1);
+        } else {
+            assert_eq!(workers.len(), min_expected_workers);
+        }
+    }
```

Comment:
Wasn't too hard to run this to a Shuttle test as well:

```rust
    #[cfg(feature = "shuttle")]
    mod shuttle_tests {
        use shuttle::rand::Rng;
        use shuttle::{check_pct, check_random};

        fn test_worker_pool_helper() {
            let mut rng = shuttle::rand::thread_rng();
            let num_worker_threads = rng.gen_range(1..=8);
            let num_concurrent_messages = rng.gen_range(1..=16);
            super::test_worker_pool(num_worker_threads, num_concurrent_messages);
        }

        #[test]
        fn test_worker_pool() {
            check_random(test_worker_pool_helper, 10000);
            check_pct(test_worker_pool_helper, 10000, 3);
        }
    }
```

Reply:
Thanks! Added together with the new test.